### ISSUE TITLE
Sudden 'child exited' and 'urgent I/O condition' after upgrading from envconsul 0.11.0 to 0.12.1

### ISSUE THREAD

**Issue by rubyalwaystaken** [hashicorp/envconsul#283]
### Envconsul version
v0.12.1

### Configuration
```hcl
"vault" {
  address = "ADDRESS"
  token = "TOKEN"
  renew_token = true
  ssl {
    enabled = true
    verify = true
	ca_cert = "CERT"
  }
  retry {
    enabled = true
    attempts = 12
    backoff = "250ms"
    max_backoff = "1m"
  }
}
"secret" {
    "path" = "mongodb"
    "no_prefix" = true
    "format" = ""
}
```

### Command
```shell
./envconsul -config hcl.config "./connection.sh"
```
connection.sh is a script which retrieves the 'username:password' data for a mongodb connection and connects to a MongoDB instance via 'mongo mongodb://username:password' with the purpose of having a CLI to the MongoDB

### Debug output
```
2022-06-01T11:07:31.711Z [INFO]  v0.12.1 (265f933)
2022-06-01T11:07:31.711Z [INFO] (runner) creating new runner (once: true)
2022-06-01T11:07:31.711Z [DEBUG] (runner) final config: {"Consul":{"Address":"","Namespace":"","Auth":{"Enabled":false,"Username":"","Password":""},"Retry":{"Attempts":12,"Backoff":250000000,"MaxBackoff":60000000000,"Enabled":true},"SSL":{"CaCert":"","CaPath":"","Cert":"","Enabled":false,"Key":"","ServerName":"","Verify":true},"Token":"","Transport":{"DialKeepAlive":30000000000,"DialTimeout":30000000000,"DisableKeepAlives":false,"IdleConnTimeout":90000000000,"MaxIdleConns":100,"MaxIdleConnsPerHost":9,"TLSHandshakeTimeout":10000000000}},"Exec":{"Command":"./connection.sh","Enabled":true,"Env":{"Denylist":[],"Custom":[],"Pristine":false,"Allowlist":[]},"KillSignal":2,"KillTimeout":30000000000,"ReloadSignal":null,"Splay":0,"Timeout":0},"KillSignal":2,"LogLevel":"debug","MaxStale":2000000000,"PidFile":"","Prefixes":[],"Pristine":false,"ReloadSignal":1,"Sanitize":false,"Secrets":[{"Format":"","NoPrefix":true,"Path":"mongodb","Keys":null}],"Services":[],"Syslog":{"Enabled":false,"Facility":"LOCAL0","Name":"consul-template"},"Upcase":false,"Vault":{"Address":"ADDRESS","Enabled":true,"Namespace":"","RenewToken":true,"Retry":{"Attempts":12,"Backoff":250000000,"MaxBackoff":60000000000,"Enabled":true},"SSL":{"CaCert":"CERT","CaPath":"","Cert":"","Enabled":true,"Key":"","ServerName":"","Verify":true},"Transport":{"DialKeepAlive":30000000000,"DialTimeout":30000000000,"DisableKeepAlives":false,"IdleConnTimeout":90000000000,"MaxIdleConns":100,"MaxIdleConnsPerHost":9,"TLSHandshakeTimeout":10000000000},"UnwrapToken":false,"DefaultLeaseDuration":300000000000},"Wait":{"Enabled":false,"Min":0,"Max":0}}
2022-06-01T11:07:31.711Z [INFO] (runner) creating watcher
2022-06-01T11:07:31.711Z [DEBUG] (watcher) adding vault.token
2022-06-01T11:07:31.711Z [INFO] looking at vault mongodb
2022-06-01T11:07:31.712Z [INFO] (runner) starting
2022-06-01T11:07:31.712Z [DEBUG] (watcher) adding vault.read(mongodb)
2022-06-01T11:07:31.826Z [DEBUG] (runner) receiving dependency vault.read(mongodb)
2022-06-01T11:07:31.826Z [INFO] (runner) running
2022-06-01T11:07:31.826Z [DEBUG] (runner) setting MONGODB_HOST from vault.read(mongodb)
2022-06-01T11:07:31.826Z [DEBUG] (cli) receiving signal "urgent I/O condition"
2022-06-01T11:07:31.826Z [DEBUG] (cli) receiving signal "urgent I/O condition"
2022-06-01T11:07:31.826Z [INFO] (child) spawning: ./connection.sh
MongoDB shell version v4.0.28
connecting to: mongodb ...
[Here is the successful connection information]
2022-06-01T11:07:32.040Z [DEBUG] (cli) receiving signal "child exited"
2022-06-01T11:07:32.040Z [DEBUG] (cli) receiving signal "urgent I/O condition"
```

### Expected behavior
Not receiving an "urgent I/O condition" and being able to use the successfully connected CLI.
This works with envconsul 0.11.0!

### Actual behavior
Receiving an "urgent I/O condition" and not being able to use the successfully connected CLI, unable to do anything. Shell seems to be stuck in a loop.

### Steps to reproduce
1. Setup Config
2. Run envconsul command with mongodb connection

### References
Not that I'm aware

**Comment by eikenb** [comment]
Hey @rubyalwaystaken, thanks for the ticket!

I'm starting work on envconsul v0.13.0 and will look at this bug as part of that process.

The "urgent I/O condition" errors come from an update to the Go runtime that uses those signals in the scheduler. We've updated consul-template to stop logging these as errors and will look into doing the same here.

The line log line `[DEBUG] (cli) receiving signal "child exited"` usually indicates that the child process was killed by an external actor, like killing it manually from the CLI (how I reproduced the error). Could an OOM or some other issue be at work here?

I'll look into this more as part of the work for this release but if you can think of anything else related to what might be killing the process envconsul is supposed to be managing please add it to the ticket. Thanks!

**Comment by rubyalwaystaken** [comment]
I'm really not sure myself as to where the 'child exited' message could come from. It does not appear when using envconsul 0.11.0 however, so it might be connected to the Go runtime change?

**Comment by eikenb** [comment]
A couple things. Could you try 0.12.0... the just to bisect a touch to see if the trigger change was in 0.12.0 or 0.12.1?

Also would it be possible for you post the contents of `connection.sh`? Maybe I'd spot something in that which might give me a clue. 

Thanks.

**Comment by rubyalwaystaken** [comment]
I tested it out with the 0.12.1 and now 0.12.0 versions from your release page and both don't work. To doublecheck I pulled a fresh 0.11.0 version and it works. 

Our process of calling mongo actually works via chaining 2 scripts.
First one is a 'mongo.sh' script which wraps the envconsul command:
```shell
#!/bin/bash

FILE=[REDACTED]/envconsul

if [[ -f "$FILE" ]]; then
    ."$FILE" -config hcl.config --app-secrets "mongodb" --cmd "./connection.sh"
fi
```

connection.sh then looks like the following:
```shell
#!/bin/bash

if [[ ! -z "${MONGODB_HOST}" ]]; then
        mongo mongodb://"${MONGODB_HOST}"/?[REDACTED]
fi
```

I tried out skipping the entire mongo.sh script and calling the command specified in mongo.sh directly (assuming that spawning a child process from another might mess with things) but it leads to the same error.

**Comment by eikenb** [comment]
Hey @rubyalwaystaken, I'm back looking at this again and have a could asks if you have time.

Are you using the envconsul docker image for this? The docker image was redone as part of 0.12.0.

Also maybe try again with a build off main if you can. *The* big dependencies of envconsul (consul-template) had an update in how it handles executing the managed process (eg. mongodb). How that works has stabilized a bit more and maybe that was fixed already there.

I looked through a diff between v0.11.0 and v0.12.0 and nothing seemed like it could cause this issue except for the dependency changes. Given how envconsul builds on consul-template that one is the primary suspect.

If you have any other thoughts/ideas please share. :)

Thanks!

**Comment by eikenb** [comment]
Hey @rubyalwaystaken,

So I've merged the fix for #305 into main and was hoping, if you have a moment, you might see if it fixes your case. I'm hopeful it will as the errors were near identical and the cause makes sense for your issue as well. If you get a chance please let me know. Thanks.

**Comment by rubyalwaystaken** [comment]
Hey @eikenb,
I just tried it out and the 0.13.0 version fixes my issue. Thanks a lot for taking the time!

**Comment by eikenb** [comment]
Great. Thanks for the followup!

Fixed by #306
